File: binary_tree_max_path_sum.py

class Node(object):
    def __init__(self,val):
        self.val = val
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
    def maxPathSum_helper(self, root):
        if root is None:
            return (0,-1*2147483648)
        l = self.maxPathSum_helper(root.left)
        r = self.maxPathSum_helper(root.right)
        m = max(l[1],r[1],l[0]+root.val,r[0]+root.val,l[0]+r[0]+root.val,root.val)
        return (max(root.val,max(l[0],r[0])+root.val),m)
    
    def maxPathSum(self,root):
        return self.maxPathSum_helper(root)[1] if root is not None else 0

if __name__ == "__main__":
    n1 = Node(1)
    n2 = Node(2)
    n3 = Node(3)
    n1.left = n2
    n1.right = n3
    print Solution().maxPathSum(n1)

File: 5/binary_tree_max_path_sum.cpp
/*
Given a binary tree, find the maximum path sum.

The path may start and end at any node in the tree.

For example:
Given the below binary tree,
       1
      / \
     2   3
Return 6.
*/

#include "header.h"

void search(TreeNode* root, int& res, int& m){
    if(!root->left && !root->right){
        res=m=root->val;
        return;
    }
    int l_m, r_m, l_res, r_res;
    l_m=r_m=l_res=r_res=INT_MIN;
    if(root->left) search(root->left, l_res, l_m);
    if(root->right) search(root->right, r_res, r_m);
    int x = max(l_m,r_m);
    m = max(x,x+root->val);
    m = max(m, root->val);
    int child_res=max(l_res,r_res);
    int res_with_curr = max(max(l_m, r_m),(l_m==INT_MIN||r_m==INT_MIN?INT_MIN:l_m+r_m))+root->val;
    res_with_curr = max(res_with_curr,root->val);
    res = max(child_res,res_with_curr);
}

int maxPathSum(TreeNode* root){
    if(!root) return 0;
    int res = 0, m=0;
    search(root, res, m);
    return res;
}

int main(){
    //string arr[] = {"-1","5","#","4","#","#","2","-4","#"};
    string arr[] = {"2","-1"};
    vector<string> v(arr,arr+sizeof(arr)/sizeof(arr[0]));
    TreeNode* root = create_tree(v);
    cout<<maxPathSum(root)<<endl;
    return 0;
}
